src/Data/Bits/Extras.hs

```diff
@@ -38,22 +43,57 @@ import Data.Word
 import Foreign.Ptr
 import Foreign.Storable
 import GHC.Base
+#if __GLASGOW_HASKELL__ >= 900
+import Control.Exception
+import GHC.Integer.Logarithms
+#endif
 
 -- $setup
 -- >>> import Data.Bits (Bits(..))
 -- >>> import Data.Word (Word)
 
--- TODO: generalize to 64 bits, etc.
+-- | Calculate the integer base 2 logarithm. The argument must be strictly
+-- positive.
+#if __GLASGOW_HASKELL__ >= 900
+log2 :: Integral a => a -> Int
+log2 = integerLog2 . toInteger
+{-# NOINLINE [1] log2 #-}
+{-# RULES
+"log2/Integer->Int" log2 = integerLog2 :: Integer -> Int
+"log2/Word->Int" log2 = wordLog2 :: Word -> Int
+"log2/Word32->Int" log2 = word32Log2 :: Word32 -> Int
+  #-}
+
+integerLog2 :: Integer -> Int
+integerLog2 n
+  | n > 0 = I# (integerLog2# n)
+  | otherwise = throw Overflow
+{-# INLINE integerLog2 #-}
+
+wordLog2 :: Word -> Int
+wordLog2 (W# n) = I# (wordLog2# n)
+{-# INLINE wordLog2 #-}
+
+word32Log2 :: Word32 -> Int
+word32Log2 !n0 =
+#else
```

Comment:
I find this bit of CPP confusing to read, since the `#if` block ends with a partial definition of `word32Log2`, but the `#else` block ends with a partial definition of `log2`, which is a different function.

Why not just define `word32Log2` on all versions of GHC, and then conditionally define `log2` in terms of it? Something like this, to be specific:

```hs
word32Log2 :: Word32 -> Int
word32Log2 !n0 =
  fromIntegral $ go (unsafeShiftR (n5 * 0x7C4ACDD) 27)
 where
  go :: Word32 -> Word8
  go !i = inlinePerformIO $ peekElemOff debruijn_log32 (fromIntegral i)
  !n1 = n0 .|. unsafeShiftR n0 1
  !n2 = n1 .|. unsafeShiftR n1 2
  !n3 = n2 .|. unsafeShiftR n2 4
  !n4 = n3 .|. unsafeShiftR n3 8
  !n5 = n4 .|. unsafeShiftR n4 16
{-# INLINE word32Log2 #-}

#if __GLASGOW_HASKELL__ >= 900
log2 = integerLog2 . toInteger
#else
log2 = word32Log2
#endif
{-# INLINE log2 #-}
```